riscv_executor/src/lib.rs

```diff
@@ -0,0 +1,776 @@
+//! A specialized executor for our RISC-V assembly that can speedup witgen.
+//!
+//! WARNING: the general witness generation/execution code over the polynomial
+//! constraints try to ensure the determinism of the instructions. If we bypass
+//! much of witness generation using the present module, we lose the
+//! non-determinism verification.
+//!
+//! TODO: perform determinism verification for each instruction independently
+//! from execution.
+
+use std::{
+    collections::HashMap,
+    io::{self, Write},
+};
+
+use ast::{
+    asm_analysis::{AnalysisASMFile, CallableSymbol, FunctionStatement, LabelStatement, Machine},
+    parsed::{asm::AssignmentRegister, Expression},
+};
+use builder::{MemoryBuilder, TraceBuilder};
+use number::{BigInt, FieldElement};
+
+#[derive(Clone, Copy)]
+pub union Elem {
```

Comment:
why not an enum?

Reply:
Because the same binary string must be interpreted as either unsigned or signed, depending on the context.

Reply:
ah I see